# Introduction to Deep Learning and Neural Network with Keras

**Final work**

**Ing. José Miguel Pereira Pontón**


# Instructions - Part A:
A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. you can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. you can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

# Instructions - Part B:
6. Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

# Instructions - Part C:

**Increate the number of epochs (5 marks)**

7. Repeat Part B but use 100 epochs this time for training.

# Instructions - Part D

**Increase the number of hidden layers (5 marks)**

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

# Preparing the environment

Import some libraries 

In [1]:
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

import pandas as pd
import numpy as np

Read the data set

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()



Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Split dataset into inputs(predictors) and y (target)


In [3]:
# First: Split dataset into input (x) and output (y) variables
# Obtain the columns
concrete_data_columns = concrete_data.columns
# Selecting all columns except Strength
x = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] 
# Strength column
y = concrete_data['Strength']

In [4]:
# Lets do some review of information of predictors
x.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [5]:
# Lets do some review of information of target
y.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [6]:
# Finally normalizing the data by substracting the mean and dividing by the standard deviation
x_norm = (x-x.mean())/x.std()

In [7]:
# Number of predictors
n_cols = x.shape[1]

Using the model from Part A, but with a modification on the hidden layer

We use 3 hidden layers of 10 nodes and and ReLU Activation Function.

In [8]:
# Defining a function for regression model
def regression_model():
    # create model
    model = Sequential()
    # First layer
    model.add(Dense(7, activation='relu',input_shape=(n_cols,)))
    # Hiden Layer
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # Compile model
    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

model = regression_model()

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [9]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_norm, y, test_size=0.3, random_state=23)
# let's have a look at the shape of the predictors set
x_train.shape

(721, 8)

2. Train the model on the training data using 50 epochs.

In [10]:
model.fit(x_train, y_train, epochs=50)

Epoch 1/50
23/23 [==============================] - 0s 607us/step - loss: 1535.3726
Epoch 2/50
23/23 [==============================] - 0s 477us/step - loss: 1519.7361
Epoch 3/50
23/23 [==============================] - 0s 475us/step - loss: 1496.8033
Epoch 4/50
23/23 [==============================] - 0s 607us/step - loss: 1455.9669
Epoch 5/50
23/23 [==============================] - 0s 607us/step - loss: 1380.1123
Epoch 6/50
23/23 [==============================] - 0s 630us/step - loss: 1243.7203
Epoch 7/50
23/23 [==============================] - 0s 650us/step - loss: 1024.6709
Epoch 8/50
23/23 [==============================] - 0s 607us/step - loss: 717.0242
Epoch 9/50
23/23 [==============================] - 0s 607us/step - loss: 424.5437
Epoch 10/50
23/23 [==============================] - 0s 651us/step - loss: 285.3130
Epoch 11/50
23/23 [==============================] - 0s 694us/step - loss: 253.6864
Epoch 12/50
23/23 [==============================] - 0s 648us/step - loss: 237

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. you can use the mean_squared_error function from Scikit-learn.


In [11]:
y_pred = model.predict(x_test)
from sklearn.metrics import mean_squared_error
print('Mean squared error on test data is %.3f' % (mean_squared_error(y_test, y_pred)))

Mean squared error on test data is 117.858


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.
5. Report the mean and the standard deviation of the mean squared errors.

In [12]:
mse = []
for i in range(50):
    model = regression_model()
    model.fit(x_train, y_train, epochs=50)
    y_pred = model.predict(x_test)
    mse.append(mean_squared_error(y_test, y_pred))

Epoch 1/50
23/23 [==============================] - 0s 607us/step - loss: 1550.1378
Epoch 2/50
23/23 [==============================] - 0s 434us/step - loss: 1521.2924
Epoch 3/50
23/23 [==============================] - 0s 504us/step - loss: 1486.1775
Epoch 4/50
23/23 [==============================] - 0s 684us/step - loss: 1433.2742
Epoch 5/50
23/23 [==============================] - 0s 603us/step - loss: 1349.7233
Epoch 6/50
23/23 [==============================] - 0s 607us/step - loss: 1218.1007
Epoch 7/50
23/23 [==============================] - 0s 607us/step - loss: 1012.7372
Epoch 8/50
23/23 [==============================] - 0s 564us/step - loss: 742.7896
Epoch 9/50
23/23 [==============================] - 0s 607us/step - loss: 480.9917
Epoch 10/50
23/23 [==============================] - 0s 650us/step - loss: 330.6269
Epoch 11/50
23/23 [==============================] - 0s 564us/step - loss: 266.3057
Epoch 12/50
23/23 [==============================] - 0s 646us/step - loss: 229

23/23 [==============================] - 0s 693us/step - loss: 98.3255
Epoch 49/50
23/23 [==============================] - 0s 520us/step - loss: 97.4981
Epoch 50/50
23/23 [==============================] - 0s 520us/step - loss: 95.6994
Epoch 1/50
23/23 [==============================] - 0s 607us/step - loss: 1535.7039
Epoch 2/50
23/23 [==============================] - 0s 476us/step - loss: 1523.3478
Epoch 3/50
23/23 [==============================] - 0s 430us/step - loss: 1499.3341
Epoch 4/50
23/23 [==============================] - 0s 434us/step - loss: 1439.2646
Epoch 5/50
23/23 [==============================] - 0s 433us/step - loss: 1320.3591
Epoch 6/50
23/23 [==============================] - 0s 477us/step - loss: 1112.4081
Epoch 7/50
23/23 [==============================] - 0s 477us/step - loss: 807.2331
Epoch 8/50
23/23 [==============================] - 0s 477us/step - loss: 514.6968
Epoch 9/50
23/23 [==============================] - 0s 477us/step - loss: 369.0797
Epoch 10/5

23/23 [==============================] - 0s 512us/step - loss: 136.9070
Epoch 46/50
23/23 [==============================] - 0s 512us/step - loss: 136.1279
Epoch 47/50
23/23 [==============================] - 0s 476us/step - loss: 135.0500
Epoch 48/50
23/23 [==============================] - 0s 475us/step - loss: 134.6903
Epoch 49/50
23/23 [==============================] - 0s 559us/step - loss: 133.5644
Epoch 50/50
23/23 [==============================] - 0s 520us/step - loss: 132.9081
Epoch 1/50
23/23 [==============================] - 0s 564us/step - loss: 1520.3358
Epoch 2/50
23/23 [==============================] - 0s 520us/step - loss: 1481.3868
Epoch 3/50
23/23 [==============================] - 0s 477us/step - loss: 1415.6227
Epoch 4/50
23/23 [==============================] - 0s 434us/step - loss: 1292.0670
Epoch 5/50
23/23 [==============================] - 0s 434us/step - loss: 1076.2242
Epoch 6/50
23/23 [==============================] - 0s 431us/step - loss: 786.1812
Epoch

23/23 [==============================] - 0s 477us/step - loss: 125.5323
Epoch 43/50
23/23 [==============================] - 0s 520us/step - loss: 123.2898
Epoch 44/50
23/23 [==============================] - 0s 520us/step - loss: 121.8659
Epoch 45/50
23/23 [==============================] - 0s 477us/step - loss: 118.7327
Epoch 46/50
23/23 [==============================] - 0s 477us/step - loss: 116.5454
Epoch 47/50
23/23 [==============================] - 0s 520us/step - loss: 114.4855
Epoch 48/50
23/23 [==============================] - 0s 477us/step - loss: 112.7134
Epoch 49/50
23/23 [==============================] - 0s 476us/step - loss: 110.9807
Epoch 50/50
23/23 [==============================] - 0s 477us/step - loss: 108.4566
Epoch 1/50
23/23 [==============================] - 0s 564us/step - loss: 1537.5969
Epoch 2/50
23/23 [==============================] - 0s 434us/step - loss: 1507.6261
Epoch 3/50
23/23 [==============================] - 0s 434us/step - loss: 1463.1165
Epoc

23/23 [==============================] - 0s 953us/step - loss: 134.9408
Epoch 40/50
23/23 [==============================] - 0s 521us/step - loss: 133.3201
Epoch 41/50
23/23 [==============================] - 0s 520us/step - loss: 132.5791
Epoch 42/50
23/23 [==============================] - 0s 473us/step - loss: 131.8611
Epoch 43/50
23/23 [==============================] - 0s 519us/step - loss: 131.2454
Epoch 44/50
23/23 [==============================] - 0s 511us/step - loss: 130.2823
Epoch 45/50
23/23 [==============================] - 0s 477us/step - loss: 129.8067
Epoch 46/50
23/23 [==============================] - 0s 477us/step - loss: 128.6395
Epoch 47/50
23/23 [==============================] - 0s 520us/step - loss: 128.6253
Epoch 48/50
23/23 [==============================] - 0s 564us/step - loss: 127.7433
Epoch 49/50
23/23 [==============================] - 0s 607us/step - loss: 127.0171
Epoch 50/50
23/23 [==============================] - 0s 737us/step - loss: 126.6858
Epoc

23/23 [==============================] - 0s 1ms/step - loss: 137.3662
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 135.9467
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 134.5554
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 133.2641
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 132.7507
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 131.3077
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 129.6694
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 128.5231
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 127.1039
Epoch 46/50
23/23 [==============================] - 0s 866us/step - loss: 126.1621
Epoch 47/50
23/23 [==============================] - 0s 564us/step - loss: 124.5886
Epoch 48/50
23/23 [==============================] - 0s 519us/step - loss: 123.9692
Epoch 49/50
23/23 [===

Epoch 35/50
23/23 [==============================] - 0s 694us/step - loss: 99.0589
Epoch 36/50
23/23 [==============================] - 0s 556us/step - loss: 96.0328
Epoch 37/50
23/23 [==============================] - 0s 608us/step - loss: 91.9130
Epoch 38/50
23/23 [==============================] - 0s 606us/step - loss: 89.2859
Epoch 39/50
23/23 [==============================] - 0s 553us/step - loss: 86.0941
Epoch 40/50
23/23 [==============================] - 0s 644us/step - loss: 83.5262
Epoch 41/50
23/23 [==============================] - 0s 563us/step - loss: 81.0822
Epoch 42/50
23/23 [==============================] - 0s 561us/step - loss: 78.6805
Epoch 43/50
23/23 [==============================] - 0s 607us/step - loss: 76.7064
Epoch 44/50
23/23 [==============================] - 0s 607us/step - loss: 74.7071
Epoch 45/50
23/23 [==============================] - 0s 602us/step - loss: 72.9218
Epoch 46/50
23/23 [==============================] - 0s 560us/step - loss: 71.0830
Epoc

23/23 [==============================] - 0s 606us/step - loss: 147.0462
Epoch 33/50
23/23 [==============================] - 0s 649us/step - loss: 145.9177
Epoch 34/50
23/23 [==============================] - 0s 606us/step - loss: 144.2025
Epoch 35/50
23/23 [==============================] - 0s 607us/step - loss: 142.5424
Epoch 36/50
23/23 [==============================] - 0s 646us/step - loss: 141.0065
Epoch 37/50
23/23 [==============================] - 0s 563us/step - loss: 140.0579
Epoch 38/50
23/23 [==============================] - 0s 607us/step - loss: 138.3401
Epoch 39/50
23/23 [==============================] - 0s 606us/step - loss: 137.3848
Epoch 40/50
23/23 [==============================] - 0s 607us/step - loss: 135.7603
Epoch 41/50
23/23 [==============================] - 0s 564us/step - loss: 134.6051
Epoch 42/50
23/23 [==============================] - 0s 607us/step - loss: 133.2902
Epoch 43/50
23/23 [==============================] - 0s 607us/step - loss: 131.9220
Epoc

Epoch 30/50
23/23 [==============================] - 0s 650us/step - loss: 152.4548
Epoch 31/50
23/23 [==============================] - 0s 649us/step - loss: 151.4014
Epoch 32/50
23/23 [==============================] - 0s 649us/step - loss: 149.9386
Epoch 33/50
23/23 [==============================] - 0s 603us/step - loss: 148.8506
Epoch 34/50
23/23 [==============================] - 0s 607us/step - loss: 147.6394
Epoch 35/50
23/23 [==============================] - 0s 564us/step - loss: 147.3329
Epoch 36/50
23/23 [==============================] - 0s 562us/step - loss: 146.1002
Epoch 37/50
23/23 [==============================] - 0s 607us/step - loss: 144.9721
Epoch 38/50
23/23 [==============================] - 0s 607us/step - loss: 143.7458
Epoch 39/50
23/23 [==============================] - 0s 564us/step - loss: 143.2738
Epoch 40/50
23/23 [==============================] - 0s 607us/step - loss: 142.2402
Epoch 41/50
23/23 [==============================] - 0s 563us/step - loss: 1

23/23 [==============================] - 0s 690us/step - loss: 154.3788
Epoch 28/50
23/23 [==============================] - 0s 650us/step - loss: 152.5738
Epoch 29/50
23/23 [==============================] - 0s 651us/step - loss: 150.0785
Epoch 30/50
23/23 [==============================] - 0s 607us/step - loss: 148.5024
Epoch 31/50
23/23 [==============================] - 0s 650us/step - loss: 145.9877
Epoch 32/50
23/23 [==============================] - 0s 607us/step - loss: 144.3980
Epoch 33/50
23/23 [==============================] - 0s 650us/step - loss: 142.1078
Epoch 34/50
23/23 [==============================] - 0s 607us/step - loss: 139.8924
Epoch 35/50
23/23 [==============================] - 0s 605us/step - loss: 138.1775
Epoch 36/50
23/23 [==============================] - 0s 606us/step - loss: 137.2096
Epoch 37/50
23/23 [==============================] - 0s 561us/step - loss: 135.0018
Epoch 38/50
23/23 [==============================] - 0s 604us/step - loss: 132.6756
Epoc

23/23 [==============================] - 0s 1ms/step - loss: 178.7689
Epoch 25/50
23/23 [==============================] - 0s 997us/step - loss: 176.0753
Epoch 26/50
23/23 [==============================] - 0s 794us/step - loss: 174.3148
Epoch 27/50
23/23 [==============================] - 0s 694us/step - loss: 169.8904
Epoch 28/50
23/23 [==============================] - 0s 694us/step - loss: 166.5135
Epoch 29/50
23/23 [==============================] - 0s 607us/step - loss: 163.9639
Epoch 30/50
23/23 [==============================] - 0s 650us/step - loss: 161.7751
Epoch 31/50
23/23 [==============================] - 0s 649us/step - loss: 159.5440
Epoch 32/50
23/23 [==============================] - 0s 694us/step - loss: 157.7855
Epoch 33/50
23/23 [==============================] - 0s 650us/step - loss: 155.5839
Epoch 34/50
23/23 [==============================] - 0s 650us/step - loss: 153.6960
Epoch 35/50
23/23 [==============================] - 0s 650us/step - loss: 152.1929
Epoch 

23/23 [==============================] - 0s 692us/step - loss: 170.8437
Epoch 22/50
23/23 [==============================] - 0s 650us/step - loss: 167.6026
Epoch 23/50
23/23 [==============================] - 0s 564us/step - loss: 164.1132
Epoch 24/50
23/23 [==============================] - 0s 591us/step - loss: 161.7440
Epoch 25/50
23/23 [==============================] - 0s 649us/step - loss: 159.3258
Epoch 26/50
23/23 [==============================] - 0s 606us/step - loss: 157.4854
Epoch 27/50
23/23 [==============================] - 0s 603us/step - loss: 156.1592
Epoch 28/50
23/23 [==============================] - 0s 607us/step - loss: 153.6913
Epoch 29/50
23/23 [==============================] - 0s 607us/step - loss: 151.8551
Epoch 30/50
23/23 [==============================] - 0s 564us/step - loss: 150.5366
Epoch 31/50
23/23 [==============================] - 0s 607us/step - loss: 149.1515
Epoch 32/50
23/23 [==============================] - 0s 564us/step - loss: 147.5564
Epoc

23/23 [==============================] - 0s 1ms/step - loss: 160.2539
Epoch 20/50
23/23 [==============================] - 0s 1ms/step - loss: 158.2525
Epoch 21/50
23/23 [==============================] - 0s 1ms/step - loss: 156.1763
Epoch 22/50
23/23 [==============================] - 0s 1ms/step - loss: 155.2235
Epoch 23/50
23/23 [==============================] - 0s 1ms/step - loss: 153.5431
Epoch 24/50
23/23 [==============================] - 0s 1ms/step - loss: 153.0101
Epoch 25/50
23/23 [==============================] - 0s 1ms/step - loss: 150.7885
Epoch 26/50
23/23 [==============================] - 0s 1ms/step - loss: 150.2382
Epoch 27/50
23/23 [==============================] - 0s 1ms/step - loss: 148.5471
Epoch 28/50
23/23 [==============================] - 0s 1ms/step - loss: 147.6987
Epoch 29/50
23/23 [==============================] - 0s 1ms/step - loss: 146.4914
Epoch 30/50
23/23 [==============================] - 0s 1ms/step - loss: 145.5524
Epoch 31/50
23/23 [=========

23/23 [==============================] - 0s 477us/step - loss: 180.7728
Epoch 18/50
23/23 [==============================] - 0s 520us/step - loss: 176.6948
Epoch 19/50
23/23 [==============================] - 0s 520us/step - loss: 172.8681
Epoch 20/50
23/23 [==============================] - 0s 477us/step - loss: 169.4729
Epoch 21/50
23/23 [==============================] - 0s 431us/step - loss: 166.0313
Epoch 22/50
23/23 [==============================] - 0s 477us/step - loss: 162.7788
Epoch 23/50
23/23 [==============================] - 0s 477us/step - loss: 159.8857
Epoch 24/50
23/23 [==============================] - 0s 477us/step - loss: 157.1149
Epoch 25/50
23/23 [==============================] - 0s 477us/step - loss: 154.6747
Epoch 26/50
23/23 [==============================] - 0s 434us/step - loss: 152.5569
Epoch 27/50
23/23 [==============================] - 0s 477us/step - loss: 150.4029
Epoch 28/50
23/23 [==============================] - 0s 434us/step - loss: 147.7929
Epoc

23/23 [==============================] - 0s 737us/step - loss: 214.2054
Epoch 15/50
23/23 [==============================] - 0s 602us/step - loss: 204.2213
Epoch 16/50
23/23 [==============================] - 0s 564us/step - loss: 196.9032
Epoch 17/50
23/23 [==============================] - 0s 564us/step - loss: 190.2305
Epoch 18/50
23/23 [==============================] - 0s 562us/step - loss: 185.6192
Epoch 19/50
23/23 [==============================] - 0s 607us/step - loss: 180.9167
Epoch 20/50
23/23 [==============================] - 0s 601us/step - loss: 176.8212
Epoch 21/50
23/23 [==============================] - 0s 607us/step - loss: 173.0843
Epoch 22/50
23/23 [==============================] - 0s 564us/step - loss: 168.5339
Epoch 23/50
23/23 [==============================] - 0s 607us/step - loss: 164.9807
Epoch 24/50
23/23 [==============================] - 0s 564us/step - loss: 161.4314
Epoch 25/50
23/23 [==============================] - 0s 563us/step - loss: 158.0871
Epoc

23/23 [==============================] - 0s 564us/step - loss: 344.0907
Epoch 12/50
23/23 [==============================] - 0s 650us/step - loss: 283.2645
Epoch 13/50
23/23 [==============================] - 0s 604us/step - loss: 241.3485
Epoch 14/50
23/23 [==============================] - 0s 606us/step - loss: 213.2720
Epoch 15/50
23/23 [==============================] - 0s 606us/step - loss: 197.7062
Epoch 16/50
23/23 [==============================] - 0s 607us/step - loss: 187.2780
Epoch 17/50
23/23 [==============================] - 0s 600us/step - loss: 179.6271
Epoch 18/50
23/23 [==============================] - 0s 607us/step - loss: 174.4494
Epoch 19/50
23/23 [==============================] - 0s 608us/step - loss: 170.3033
Epoch 20/50
23/23 [==============================] - 0s 606us/step - loss: 166.1924
Epoch 21/50
23/23 [==============================] - 0s 564us/step - loss: 162.8491
Epoch 22/50
23/23 [==============================] - 0s 607us/step - loss: 159.9885
Epoc

23/23 [==============================] - 0s 564us/step - loss: 722.3918
Epoch 9/50
23/23 [==============================] - 0s 520us/step - loss: 496.0725
Epoch 10/50
23/23 [==============================] - 0s 520us/step - loss: 373.0448
Epoch 11/50
23/23 [==============================] - 0s 519us/step - loss: 307.4402
Epoch 12/50
23/23 [==============================] - 0s 518us/step - loss: 267.7926
Epoch 13/50
23/23 [==============================] - 0s 514us/step - loss: 242.3244
Epoch 14/50
23/23 [==============================] - 0s 520us/step - loss: 227.1304
Epoch 15/50
23/23 [==============================] - 0s 477us/step - loss: 217.2654
Epoch 16/50
23/23 [==============================] - 0s 520us/step - loss: 210.2766
Epoch 17/50
23/23 [==============================] - 0s 564us/step - loss: 205.6960
Epoch 18/50
23/23 [==============================] - 0s 564us/step - loss: 201.6646
Epoch 19/50
23/23 [==============================] - 0s 562us/step - loss: 197.4400
Epoch

23/23 [==============================] - 0s 542us/step - loss: 1418.2487
Epoch 8/50
23/23 [==============================] - 0s 564us/step - loss: 1281.2555
Epoch 9/50
23/23 [==============================] - 0s 520us/step - loss: 1029.2098
Epoch 10/50
23/23 [==============================] - 0s 520us/step - loss: 689.4947
Epoch 11/50
23/23 [==============================] - 0s 476us/step - loss: 405.0877
Epoch 12/50
23/23 [==============================] - 0s 520us/step - loss: 287.1113
Epoch 13/50
23/23 [==============================] - 0s 520us/step - loss: 234.9901
Epoch 14/50
23/23 [==============================] - 0s 520us/step - loss: 211.9632
Epoch 15/50
23/23 [==============================] - 0s 477us/step - loss: 198.7767
Epoch 16/50
23/23 [==============================] - 0s 475us/step - loss: 190.6216
Epoch 17/50
23/23 [==============================] - 0s 520us/step - loss: 184.9066
Epoch 18/50
23/23 [==============================] - 0s 520us/step - loss: 179.9993
Epo

23/23 [==============================] - 0s 477us/step - loss: 1407.1515
Epoch 6/50
23/23 [==============================] - 0s 477us/step - loss: 1305.3065
Epoch 7/50
23/23 [==============================] - 0s 477us/step - loss: 1139.2262
Epoch 8/50
23/23 [==============================] - 0s 512us/step - loss: 929.4473
Epoch 9/50
23/23 [==============================] - 0s 563us/step - loss: 747.1108
Epoch 10/50
23/23 [==============================] - 0s 513us/step - loss: 633.0352
Epoch 11/50
23/23 [==============================] - 0s 563us/step - loss: 535.7568
Epoch 12/50
23/23 [==============================] - 0s 608us/step - loss: 451.1234
Epoch 13/50
23/23 [==============================] - 0s 563us/step - loss: 371.7177
Epoch 14/50
23/23 [==============================] - 0s 564us/step - loss: 308.9561
Epoch 15/50
23/23 [==============================] - 0s 520us/step - loss: 267.2889
Epoch 16/50
23/23 [==============================] - 0s 519us/step - loss: 243.9149
Epoch

23/23 [==============================] - 0s 867us/step - loss: 1507.1782
Epoch 3/50
23/23 [==============================] - 0s 564us/step - loss: 1478.2257
Epoch 4/50
23/23 [==============================] - 0s 519us/step - loss: 1425.2899
Epoch 5/50
23/23 [==============================] - 0s 520us/step - loss: 1329.4717
Epoch 6/50
23/23 [==============================] - 0s 477us/step - loss: 1162.1461
Epoch 7/50
23/23 [==============================] - 0s 477us/step - loss: 909.0381
Epoch 8/50
23/23 [==============================] - 0s 520us/step - loss: 593.2717
Epoch 9/50
23/23 [==============================] - 0s 520us/step - loss: 334.3555
Epoch 10/50
23/23 [==============================] - 0s 564us/step - loss: 244.4734
Epoch 11/50
23/23 [==============================] - 0s 520us/step - loss: 222.5350
Epoch 12/50
23/23 [==============================] - 0s 520us/step - loss: 208.9575
Epoch 13/50
23/23 [==============================] - 0s 564us/step - loss: 200.8328
Epoch 

23/23 [==============================] - 0s 737us/step - loss: 95.4438
Epoch 50/50
23/23 [==============================] - 0s 564us/step - loss: 94.3781
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1534.0540
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1520.9724
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1501.3369
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1462.6631
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1383.5676
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1215.9481
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 891.3059
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 495.2408
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 285.2663
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 232.8082
Epoch 11/50
23/23 [==========

23/23 [==============================] - 0s 520us/step - loss: 134.3433
Epoch 42/50
23/23 [==============================] - 0s 514us/step - loss: 133.9472
Epoch 43/50
23/23 [==============================] - 0s 520us/step - loss: 132.9216
Epoch 44/50
23/23 [==============================] - 0s 520us/step - loss: 132.1231
Epoch 45/50
23/23 [==============================] - 0s 477us/step - loss: 132.0809
Epoch 46/50
23/23 [==============================] - 0s 510us/step - loss: 130.5356
Epoch 47/50
23/23 [==============================] - 0s 477us/step - loss: 130.1448
Epoch 48/50
23/23 [==============================] - 0s 473us/step - loss: 129.3283
Epoch 49/50
23/23 [==============================] - 0s 520us/step - loss: 129.2691
Epoch 50/50
23/23 [==============================] - 0s 562us/step - loss: 128.7966
Epoch 1/50
23/23 [==============================] - 0s 564us/step - loss: 1542.9885
Epoch 2/50
23/23 [==============================] - 0s 473us/step - loss: 1533.9910
Epoc

23/23 [==============================] - 0s 607us/step - loss: 96.1464
Epoch 39/50
23/23 [==============================] - 0s 604us/step - loss: 93.6676
Epoch 40/50
23/23 [==============================] - 0s 607us/step - loss: 91.4328
Epoch 41/50
23/23 [==============================] - 0s 650us/step - loss: 89.9015
Epoch 42/50
23/23 [==============================] - 0s 649us/step - loss: 88.2730
Epoch 43/50
23/23 [==============================] - 0s 650us/step - loss: 85.9033
Epoch 44/50
23/23 [==============================] - 0s 607us/step - loss: 84.7669
Epoch 45/50
23/23 [==============================] - 0s 650us/step - loss: 82.9660
Epoch 46/50
23/23 [==============================] - 0s 607us/step - loss: 81.3833
Epoch 47/50
23/23 [==============================] - 0s 607us/step - loss: 80.0988
Epoch 48/50
23/23 [==============================] - 0s 606us/step - loss: 78.7854
Epoch 49/50
23/23 [==============================] - 0s 605us/step - loss: 77.5300
Epoch 50/50
23/2

23/23 [==============================] - 0s 520us/step - loss: 135.0515
Epoch 38/50
23/23 [==============================] - 0s 515us/step - loss: 133.1463
Epoch 39/50
23/23 [==============================] - 0s 521us/step - loss: 131.9063
Epoch 40/50
23/23 [==============================] - 0s 520us/step - loss: 130.6002
Epoch 41/50
23/23 [==============================] - 0s 564us/step - loss: 130.0269
Epoch 42/50
23/23 [==============================] - 0s 520us/step - loss: 128.2837
Epoch 43/50
23/23 [==============================] - 0s 824us/step - loss: 127.7946
Epoch 44/50
23/23 [==============================] - 0s 682us/step - loss: 126.3728
Epoch 45/50
23/23 [==============================] - 0s 564us/step - loss: 125.5075
Epoch 46/50
23/23 [==============================] - 0s 520us/step - loss: 124.5736
Epoch 47/50
23/23 [==============================] - 0s 519us/step - loss: 123.5995
Epoch 48/50
23/23 [==============================] - 0s 477us/step - loss: 122.5937
Epoc

23/23 [==============================] - 0s 693us/step - loss: 151.9983
Epoch 32/50
23/23 [==============================] - 0s 648us/step - loss: 150.0652
Epoch 33/50
23/23 [==============================] - 0s 607us/step - loss: 147.9916
Epoch 34/50
23/23 [==============================] - 0s 600us/step - loss: 146.1929
Epoch 35/50
23/23 [==============================] - 0s 607us/step - loss: 144.0282
Epoch 36/50
23/23 [==============================] - 0s 564us/step - loss: 142.5342
Epoch 37/50
23/23 [==============================] - 0s 607us/step - loss: 141.0376
Epoch 38/50
23/23 [==============================] - 0s 606us/step - loss: 138.9513
Epoch 39/50
23/23 [==============================] - 0s 607us/step - loss: 137.6669
Epoch 40/50
23/23 [==============================] - 0s 607us/step - loss: 136.0812
Epoch 41/50
23/23 [==============================] - 0s 648us/step - loss: 134.5468
Epoch 42/50
23/23 [==============================] - 0s 650us/step - loss: 132.4298
Epoc

23/23 [==============================] - 0s 650us/step - loss: 153.7183
Epoch 29/50
23/23 [==============================] - 0s 605us/step - loss: 152.2861
Epoch 30/50
23/23 [==============================] - 0s 694us/step - loss: 150.9875
Epoch 31/50
23/23 [==============================] - 0s 781us/step - loss: 149.6666
Epoch 32/50
23/23 [==============================] - 0s 737us/step - loss: 148.4211
Epoch 33/50
23/23 [==============================] - 0s 693us/step - loss: 147.3207
Epoch 34/50
23/23 [==============================] - 0s 648us/step - loss: 147.0055
Epoch 35/50
23/23 [==============================] - 0s 650us/step - loss: 145.2908
Epoch 36/50
23/23 [==============================] - 0s 694us/step - loss: 144.4420
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 143.1431
Epoch 38/50
23/23 [==============================] - 0s 867us/step - loss: 141.9976
Epoch 39/50
23/23 [==============================] - 0s 651us/step - loss: 141.2107
Epoch 

5. Report the mean and the standard deviation of the mean squared errors.

In [13]:
mean = np.mean(mse)
mean

114.43287273930177

In [14]:
std = np.std(mse)
std

18.523912884245586

In [15]:
print("Mean squared errors for 50 regression models: mean = %.3f, std dev = %.3f" %(mean, std))

Mean squared errors for 50 regression models: mean = 114.433, std dev = 18.524


How does the mean of the mean squared errors compare to that from Step B?

In Step B, I record a MSE of 159.145 and in Part D I record a MSE of 117.858, we can notice that in Part B, we had only 1 hidden layer and in Part D we have 3 hidden layers. The number of hidden layers improve the result

And if we review the MSE of the 50 regression model of Part A, B and D, and 100 regression model of Part C:

|--- |Part A | Part B| Part C | Part D |
| --- | ---- | --- | --- | --- |
|**MSE**|203.422 | 151.077| 110.776 | 114.432 |
| **STD** | 206.667 | 10.433| 13.230 | 18.523 |

When we analyze the mean and standard deviation of multiple regression models, we can notice that in Part D, compared to Part B, the MSE was better but the STD increased. 

And if we compared to part C what was similar but with less hidden layers, and more Epochs. Part D obtained poor results, MSE was 4 points more and the STD increased, but compared to Part B, is a better model.

We can obtain better fitting for the model but the time to process the data increase too.

If I have to decided what model implement, I would choose the model in Part C.

**End of Part D.**